src/runtime/sgx/untrusted/sgx_module.cc

```diff
@@ -0,0 +1,208 @@
+      LOG(INFO) << "Could not save SGX launch token to " << token_file;
+    }
+  }
+
+  const char* type_key() const final {
+    return "sgx";
+  }
+
+  PackedFunc GetFunction(
+      const std::string& name,
+      const std::shared_ptr<ModuleNode>& sptr_to_self) final {
+    size_t func_id = std::distance(exports_.begin(),
+        std::find(exports_.begin(), exports_.end(), name));
+    CHECK_LT(func_id, exports_.size())
+      << "\"" << name << "\" is not an enclave export.";
+    if (func_id >= exports_.size()) return PackedFunc();
+    return PackedFunc([this, func_id](TVMArgs args, TVMRetValue* rv) {
+        sgx::EnclaveContext ctx(this);
+        TVM_SGX_CHECKED_CALL(tvm_ecall_packed_func(eid_, func_id,
+              args.values, args.type_codes, args.num_args, rv));
+      });
+  }
+
+  void RunWorkers(int num_tasks, void* tg) {
+    std::function<void(int)> runner = [this, tg](int _worker_id) {
+      this->GetFunction("__tvm_run_worker__",
+                        std::shared_ptr<SGXModuleNode>(nullptr))(tg);
+    };
+    thread_group_.reset(new tvm::runtime::threading::ThreadGroup(
+          num_tasks, runner, false /* include_main_thread */));
+  }
+
+  void JoinThreads() {
+    thread_group_->Join();
+  }
+
+  void RegisterExport(std::string name) {
+    exports_.push_back(name);
+  }
+
+ private:
+  // ID of the loaded enclave
+  sgx_enclave_id_t eid_;
+  // Names and IDs of functions exported by the enclave module
+  std::vector<std::string> exports_;
+  std::unique_ptr<tvm::runtime::threading::ThreadGroup> thread_group_;
+};
+
+namespace sgx {
+
+TVM_REGISTER_GLOBAL("__sgx_thread_group_launch__")
+.set_body([](TVMArgs args, TVMRetValue* rv) {
+  EnclaveContext::GetModule()->RunWorkers(args[0], args[1]);
+});
+
+TVM_REGISTER_GLOBAL("__sgx_thread_group_join__")
+.set_body([](TVMArgs args, TVMRetValue* rv) {
+  EnclaveContext::GetModule()->JoinThreads();
+});
+
+TVM_REGISTER_GLOBAL("__sgx_set_last_error__")
+.set_body([](TVMArgs args, TVMRetValue* rv) {
+  std::string err = args[0];
+  TVMAPISetLastError(err.c_str());
+});
+
+extern "C" {
+
+void tvm_ocall_register_export(const char* name) {
+  EnclaveContext::GetModule()->RegisterExport(name);
+}
+
+void tvm_ocall_packed_func(const char* name,
+                           const TVMValue* arg_values,
+                           const int* type_codes,
+                           int num_args,
+                           TVMValue* ret_val,
+                           int* ret_type_code) {
+  const PackedFunc* f = Registry::Get(name);
+  CHECK(f != nullptr) << "ocall to nonexistent function \"" << name << "\"";
+  TVMRetValue rv;
+  f->CallPacked(TVMArgs(arg_values, type_codes, num_args), &rv);
+  rv.MoveToCHost(ret_val, ret_type_code);
+}
+
+// Allocates space for return values. The returned pointer is only valid between
+// successive calls to `tvm_ocall_reserve_space`.
+void* tvm_ocall_reserve_space(size_t num_bytes) {
+  static thread_local std::vector<uint8_t> buf;
```

Comment:
Be careful, there can be memory alignment issue here. buf creates a space that only needs to be aligned per byte, we are storing TVMByteArray, which have a higher alignment requirement. The simplest way would be to change content to ```uint64_t``` and allocate accordingly for now. A better approach would be also passed in alignment as a second argument, and do proper alignment according to the alignment of head address.